### PR TITLE
add benchmark for on_initialize

### LINKED ISSUE #74: add benchmark support for on_finalize in offchain worker
**Description**
We should also support on_finalize weight got from benchmark in offchain worker pallet.

**Acceptance Criteria**
- [ ] Get the weight by run benchmark.

### PR DESCRIPTION
**Pull request by suinuj**
add dummy extrinsic for benchmark testing since don't find out how to run benchmark for hook functions.

### CODE REVIEW

**Review by Satoshi-Kusumoto — CHANGES_REQUESTED** [review]
Pls modify according to our discussion

**Comment by suinuj** [comment]
Add two parameters for ocw pallets MaximumWeightForDataAggregation and MaximumCommitsPerSession in new commit.
1. After investigate the schedule pallet, it is a good idea to control the length of data via a configuration, so MaximumCommitsPerSession added.
2. Current benchmarks can't support the weights for hook functions like on_initialize
3. The on_initialize in schedule provides an example of how to compute the weight, but it is just suitable for pallet like schedule since most of the weight is from the call, the number of scheduled tasks is very limited. So it can iterate the data on-chain and don't care about the weight used to compute the weight for on_initialize. But ocw is different, it is close to executing the data aggregation again.

**Comment by Satoshi-Kusumoto** [comment]
Pls check benchmark in substrate/timestamp

File: pallets/offchain-worker/src/lib.rs [hunk 1]
```diff
@@ -117,6 +117,14 @@ pub mod pallet {
 		type Reward: OnUnbalanced<PositiveImbalanceOf<Self>>;
 		type OcwQueryReward: Get<<<Self as Config>::Currency as Currency<<Self as frame_system::Config>::AccountId>>::Balance>;
 		type WeightInfo: weights::WeightInfo;
+
+		/// The maximum weight indicate the maximum weight used for data aggregation at the end of a session
+		type MaximumWeightForDataAggregation: Get<Weight>;
```
Comment (Satoshi-Kusumoto) [inline pallets/offchain-worker/src/lib.rs:122]: These are to cap the weight for benchmarks right? But I cann't find it's usage in the code

File: pallets/offchain-worker/src/lib.rs [hunk 2]
```diff
@@ -329,6 +340,21 @@ pub mod pallet {
...
+			let query_session_length: usize = T::QuerySessionLength::get() as usize;
+			let index_in_session = TryInto::<usize>::try_into(block_number).map_or(query_session_length, |bn| bn % query_session_length);
+			let last_block_number = query_session_length - 1;
+
+			// Clear claim at the first block of a session
+			if index_in_session == 0 {
+				Self::clear_claim();
+			// Do aggregation at last block of a session
+			} else if index_in_session == last_block_number {
+				Self::aggregate_query_result();
```
Comment (Satoshi-Kusumoto) [inline pallets/offchain-worker/src/lib.rs:354]: What does this mean? All query is sent at the end of a session?

**Review by Satoshi-Kusumoto — COMMENTED** [review]
Pls (also in future PRs), briefly write the basic designs (say: cap weight for benchmarks, have a commit counting and is cleared at session beginning. Basically these kind of things should be either in the issue or in thee PR. Because the comment you gave may be obsolete and I cannot know which part is still up to date. This also suggest that if the implementation or idea is changed a log, it's better to close the PR and start a new one

File: pallets/offchain-worker/src/lib.rs (hunk 1 above)
Comment (suinuj) [reply pallets/offchain-worker/src/lib.rs:122]: may use it and the weight from benchmark to check if the MaximumCommitsPerSession is reasonable.

File: pallets/offchain-worker/src/lib.rs (hunk 2 above)
Comment (suinuj) [reply pallets/offchain-worker/src/lib.rs:354]: query result can be submitted anytime, we just do aggregation at the end of a session.

**Review by Satoshi-Kusumoto — APPROVED** [review]
Next time it's better to have a summary of what you have done.
e.x.: moved on_finalize logic to do_finalize()
Otherwise it's very hard for me to understand the PR, especially because it was initialised a long time ago

**Comment by suinuj** [comment]
> Next time it's better to have a summary of what you have done.
> e.x.: moved on_finalize logic to do_finalize()
> Otherwise it's very hard for me to understand the PR, especially because it was initialised a long time ago

Sure, it is a special case, will pay attention to the good PR description.

**Pull request merged**